File: src/modules/Grinder/useCases/euclidean.ts
/**
 * Euclidean rhythm generator — distributes k hits across n steps as evenly as possible.
 * Bjorklund algorithm (Toussaint 2005). Used to generate patterns in the step sequencer.
 */

export function euclidean(hits: number, steps: number, rotation: number = 0): boolean[] {
    if (steps <= 0) { return []; }
    if (hits <= 0) { return Array(steps).fill(false); }
    if (hits >= steps) { return Array(steps).fill(true); }

    // Bjorklund iterative algorithm
    type Group = boolean[];
    let groups: Group[] = [];
    for (let i = 0; i < hits; i++) { groups.push([true]); }
    for (let i = 0; i < steps - hits; i++) { groups.push([false]); }

    while (true) {
        const firstGroup = groups[0];
        if (!firstGroup) { break; }

        const splitPos = groups.findIndex((g) =>
            g.length !== firstGroup.length || g.some((v, j) => v !== firstGroup[j]!)
        );

        if (splitPos <= 0 || splitPos >= groups.length) { break; }

        const remainder = groups.length - splitPos;
        const take = Math.min(remainder, splitPos);

        const newGroups: Group[] = [];
        for (let i = 0; i < splitPos; i++) {
            const combined = [...groups[i]!];
            if (i < take) {
                combined.push(...groups[splitPos + i]!);
            }
            newGroups.push(combined);
        }
        for (let i = splitPos + take; i < groups.length; i++) {
            newGroups.push(groups[i]!);
        }

        groups = newGroups;
        if (groups.length <= 1) { break; }
    }

    // Flatten
    const pattern: boolean[] = [];
    for (const group of groups) {
        pattern.push(...group);
    }
    while (pattern.length < steps) { pattern.push(false); }

    // Rotate
    if (rotation > 0 && steps > 0) {
        const rot = rotation % steps;
        return [...pattern.slice(rot), ...pattern.slice(0, rot)];
    }
    return pattern;
}

File: src-tauri/grinder/src/euclidean.rs
//! Euclidean rhythm generator — distributes k hits across n steps as evenly as possible.
//! Implements the Bjorklund algorithm (Toussaint 2005).

/// Generate a Euclidean rhythm pattern.
/// Returns a Vec<bool> of length `steps` with `hits` trues distributed as evenly as possible.
/// `rotation` shifts the pattern cyclically.
pub fn euclidean(hits: usize, steps: usize, rotation: usize) -> Vec<bool> {
    if steps == 0 { return vec![]; }
    if hits == 0 { return vec![false; steps]; }
    if hits >= steps { return vec![true; steps]; }

    let mut pattern = Vec::with_capacity(steps);

    // Start with two groups: hits x [true] and (steps-hits) x [false]
    let mut groups: Vec<Vec<bool>> = Vec::new();
    for _ in 0..hits { groups.push(vec![true]); }
    for _ in 0..(steps - hits) { groups.push(vec![false]); }

    // Iteratively distribute remainder group onto the larger group
    loop {
        // Find where the two different patterns meet
        let first = &groups[0].clone();
        let split_pos = groups.iter().position(|g| g != first);

        match split_pos {
            None => break, // all groups are identical
            Some(pos) => {
                if pos == 0 || groups.len() - pos <= 1 { break; }

                let remainder_count = groups.len() - pos;
                let take_count = remainder_count.min(pos);

                // Append remainder groups onto the end of leading groups
                let mut new_groups = Vec::new();
                for i in 0..pos {
                    let mut combined = groups[i].clone();
                    if i < take_count {
                        combined.extend_from_slice(&groups[pos + i]);
                    }
                    new_groups.push(combined);
                }
                // Any leftover remainder groups
                for i in (pos + take_count)..groups.len() {
                    new_groups.push(groups[i].clone());
                }

                groups = new_groups;
                if groups.len() <= 1 { break; }
            }
        }
    }

    // Flatten groups into pattern
    for group in &groups {
        pattern.extend_from_slice(group);
    }
    pattern.truncate(steps);

    // Rotation
    if rotation > 0 && steps > 0 {
        let rot = rotation % steps;
        let mut rotated = Vec::with_capacity(steps);
        rotated.extend_from_slice(&pattern[rot..]);
        rotated.extend_from_slice(&pattern[..rot]);
        rotated
    } else {
        pattern
    }
}
